### PR TITLE
Stub form api

### PR DESCRIPTION
**Pull request by 0atman**
Create stubs:
- [x] Create a form using the API
- [x] Read a form
- [x] Update a form
- [x] Delete a form (optional)
- [x] List forms (all forms, optional)

#### What problem does the pull request solve?

If we quickly publish a stub API, we can work to the same interface while building out functionality.

#### Checklist

- [x] I've used the pull request template
- [x] I've linked this PR to the relevant issue (if mission work)
- [x] I've written unit tests for these changes (if code change)
- [x] I've updated the documentation in (If any documentation requires updating)
    - [ ] README.md
    - [ ] Elsewhere (please link)

### CODE REVIEW

**Comment by 0atman** [comment]
@danielburnley Just checking I've understood the layout we have Services, which are controllers, and Repositories, which are Models, and the Server routes act as Views?

**Comment by danielburnley** [comment]
@0atman Sorta yea - though this isn't really MVC as such, more split by responsibilities

- `server.rb` for now contains the actual routes, which is where the logic for kinda processing the HTML requests will go (see: turning request bodies into ruby hashes) 
- `services` will handle the actual business logic, these shouldn't know _anything_ about HTML stuff and will operate on just ruby hashes - Unit tests for this can therefore not need to do anything db related for setup and can stub repositories via dependency inversion stuff
- `repositories` will be where anything database related can go, similarly to the `server.rb` anything related to databases should stay here and services shouldn't see any

At least, eventually that'll be the case. In the short term we arguably don't need the service layer if all we're doing is parsing a request and throwing it straght into the DB, kinda becomes a bit like a pointless layer in that case. 

For stubbing the API however, we can probs ignore all of that and just return the json in the `server.rb` straight away without the need for any other layers

**Comment by 0atman** [comment]
Swagger inspector looks good (run the server locally and try it here https://inspector.swagger.io/builder?url=http%3A%2F%2Flocalhost%3A9292%2Fapi%2Fv1%2Fswagger_doc)

Screenshot from 2022-05-09 11-33-32

File: lib/server.rb [hunk 1]
```diff
@@ -1,10 +1,29 @@
...
+  resource :forms do
+    desc 'Returns all forms.'
+    params do
+      requires :result, type: Integer, desc: 'Result ID.'
+    end
+    route_param :result do
+      get do
+        Services::Example.new.execute(params[:result])
+      end
+    end
```
Comment (0atman) [inline lib/server.rb:18]: @danielburnley calling this out early, like the format of grape so far? Big example looks pretty tasty to me https://github.com/ruby-grape/grape#basic-usage

**Comment by 0atman** [comment]
So far

**Comment by 0atman** [comment]
Tweaked schema based on our chat earlier
Screenshot from 2022-05-10 14-44-41

File: lib/server.rb [hunk 2]
```diff
@@ -1,10 +1,67 @@
...
+  resource :forms do
+    desc "Create a form."
+    params do
+      requires :form, type: String, desc: "Form data."
+    end
+    post do
+      {
+        form_id: 1,
+        form_name: "form name",
+        user: "user@example.com"
```
Comment (danielburnley) [inline lib/server.rb:18]: i'd probably go with:

- `id` (the form is redundant)
- `name` (same)
- `submission_email` - it'll be where we send the submissions

Comment (0atman) [reply lib/server.rb:18]: Nice, yeah that makes sense

**Comment by 0atman** [comment]
Format now

```json
{
    "id": 12,
    "name": "form name",
    "submission_email": "user@example.com"
}
```

File: lib/server.rb [hunk 3]
```diff
@@ -1,10 +1,67 @@
...
+          name: "form name",
+          submission_email: "user@example.com"
+        }
+      end
+    end
+
+    desc "Update a form."
+    params do
+      requires :id, type: String, desc: "Form ID."
+      requires :data, type: String, desc: "Form data."
```
[the remark this answers is not in the document]
Comment (0atman) [reply lib/server.rb]: cheers, fixed

File: lib/services/example.rb [hunk 4]
```diff
@@ -2,4 +2,14 @@ class Services::Example
...
+
+  def get_forms(name)
+    JSON({
+           name => {
+             item2: {
+               item3: "ok"
+             }
+           }
+         })
+  end
```
[the remark this answers is not in the document]
Comment (0atman) [reply lib/services/example.rb]: gonna be kind to github's storage and remove it, well spotted!

**Review by danielburnley — APPROVED** [review]

**Comment by 0atman** [comment]
Listing forms OK:

Screenshot from 2022-05-11 11-45-37

**Pull request merged**